File: 3108_minimum_cost_walk_in_weighted_graph/solution.php
<?php
// LeetCode 3108 - Minimum Cost Walk in Weighted Graph
// https://leetcode.com/problems/minimum-cost-walk-in-weighted-graph/

class Solution {
    public $p;
    public $size;
    function minimumCost($n, $edges, $query) {
        $this->p = range(0, $n - 1);
        $this->size = array_fill(0, $n, 1);
        $g = array_fill(0, $n, -1);
        foreach ($edges as $e) $this->unite($e[0], $e[1]);
        foreach ($edges as $e) {
            $root = $this->find($e[0]);
            $g[$root] &= $e[2];
        }
        $ans = [];
        for ($i = 0; $i < count($query); $i++) {
            $u = $query[$i][0];
            $v = $query[$i][1];
            if ($u === $v) $ans[$i] = 0;
            else {
                $a = $this->find($u);
                $b = $this->find($v);
                $ans[$i] = $a === $b ? $g[$a] : -1;
            }
        }
        return $ans;
    }
    function find($x) {
        if ($this->p[$x] !== $x) $this->p[$x] = $this->find($this->p[$x]);
        return $this->p[$x];
    }
    function unite($a, $b) {
        $pa = $this->find($a);
        $pb = $this->find($b);
        if ($pa === $pb) return;
        if ($this->size[$pa] > $this->size[$pb]) {
            $this->p[$pb] = $pa;
            $this->size[$pa] += $this->size[$pb];
        } else {
            $this->p[$pa] = $pb;
            $this->size[$pb] += $this->size[$pa];
        }
    }
}

File: 3108_minimum_cost_walk_in_weighted_graph/Solution.swift
// LeetCode 3108 - Minimum Cost Walk in Weighted Graph
// https://leetcode.com/problems/minimum-cost-walk-in-weighted-graph/

class Solution {
    func minimumCost(_ n: Int, _ edges: [[Int]], _ query: [[Int]]) -> [Int] {
        var p = Array(0..<n)
        var size = Array(repeating: 1, count: n)
        func find(_ x: Int) -> Int {
            if p[x] != x { p[x] = find(p[x]) }
            return p[x]
        }
        func unite(_ a: Int, _ b: Int) {
            var pa = find(a), pb = find(b)
            if pa == pb { return }
            if size[pa] > size[pb] {
                p[pb] = pa
                size[pa] += size[pb]
            } else {
                p[pa] = pb
                size[pb] += size[pa]
            }
        }
        for e in edges { unite(e[0], e[1]) }
        var g = Array(repeating: -1, count: n)
        for e in edges {
            g[find(e[0])] &= e[2]
        }
        return query.map { q in
            if q[0] == q[1] { return 0 }
            let a = find(q[0]), b = find(q[1])
            return a == b ? g[a] : -1
        }
    }
}
